src/main/scala/gitbucket/core/service/SystemSettingsService.scala

```diff
@@ -177,11 +179,18 @@ object SystemSettingsService {
   ) {
 
     def baseUrl(request: HttpServletRequest): String =
-      baseUrl.fold {
-        val url = request.getRequestURL.toString
-        val len = url.length - (request.getRequestURI.length - request.getContextPath.length)
-        url.substring(0, len).stripSuffix("/")
-      }(_.stripSuffix("/"))
+      baseUrl.fold(parseBaseUrl(request))(_.stripSuffix("/"))
+
+    def parseBaseUrl(req: HttpServletRequest): String = {
+      val url = req.getRequestURL.toString
+      val path = req.getRequestURI
+      val contextPath = req.getContextPath
+      val len = url.length - path.length + contextPath.length
+
+      val base = url.substring(0, len).stripSuffix("/")
+      Option(req.getHeader(XForwardedProto))
+        .fold(base)(_.toLowerCase() + base.dropWhile(_ != ':'))
```

Comment:
I think it might be dangerous to believe `X-Forward-Proto` header unconditinally.
It safer to confirm whether the value of `X-Forward-Proto` is https, then replace protocol if it's true.

Reply:
FYI: Scalatra checks if the header is `https` as https://github.com/scalatra/scalatra/blob/be9fc5119e6e34c4882d848cf86d9808acea8b68/core/src/main/scala/org/scalatra/ScalatraBase.scala#L696-L700

Reply:
@int128 as far as I can see, here we can not use Scalatra's built-in stuff. I'd rather be wrong here because I want to use framework support to handle this instead of rolling my own solution.